hoomd/mpcd/ReverseNonequilibriumShearFlow.cc

```diff
@@ -0,0 +1,431 @@
+                      detail::CompareMomentumToTarget(m_target_momentum, h_tag.data));
+            std::sort(h_particles_hi.data,
+                      h_particles_hi.data + m_num_hi,
+                      detail::CompareMomentumToTarget(-m_target_momentum, h_tag.data));
+            }
+        }
+    }
+
+/*!
+ * Stage particles momenta from both slabs into a queue for swapping
+ */
+void mpcd::ReverseNonequilibriumShearFlow::stageSwapParticles()
+    {
+    // determine number of pairs to swap
+    unsigned int num_lo_global = m_num_lo;
+    unsigned int num_hi_global = m_num_hi;
+#ifdef ENABLE_MPI
+    if (m_sysdef->isDomainDecomposed())
+        {
+        auto mpi_comm = m_exec_conf->getMPICommunicator();
+        MPI_Allreduce(MPI_IN_PLACE, &num_lo_global, 1, MPI_UNSIGNED, MPI_SUM, mpi_comm);
+        MPI_Allreduce(MPI_IN_PLACE, &num_hi_global, 1, MPI_UNSIGNED, MPI_SUM, mpi_comm);
+        }
+#endif // ENABLE_MPI
+    const unsigned int num_pairs = std::min(m_num_swap, std::min(num_lo_global, num_hi_global));
+
+    // stage swaps into queue
+    ArrayHandle<Scalar2> h_particles_lo(m_particles_lo, access_location::host, access_mode::read);
+    ArrayHandle<Scalar2> h_particles_hi(m_particles_hi, access_location::host, access_mode::read);
+    ArrayHandle<Scalar2> h_particles_staged(m_particles_staged,
+                                            access_location::host,
+                                            access_mode::overwrite);
+    m_num_staged = 0;
+    unsigned int lo_idx = 0;
+    unsigned int hi_idx = 0;
+    for (unsigned int i = 0; i < num_pairs; ++i)
+        {
+#ifdef ENABLE_MPI
+        if (m_sysdef->isDomainDecomposed())
+            {
+            const auto mpi_comm = m_exec_conf->getMPICommunicator();
+            const int rank = m_exec_conf->getRank();
+
+            // find particle in lo slab closest to +p on all ranks
+            Scalar_Int lo_swap;
+            lo_swap.i = rank;
+            if (std::isinf(m_target_momentum))
+                {
+                lo_swap.s = (lo_idx < m_num_lo) ? h_particles_lo.data[lo_idx].y
+                                                : std::numeric_limits<Scalar>::min();
+                MPI_Allreduce(MPI_IN_PLACE,
+                              &lo_swap,
+                              1,
+                              MPI_HOOMD_SCALAR_INT,
+                              MPI_MAXLOC,
+                              mpi_comm);
+                }
+            else
+                {
+                lo_swap.s = (lo_idx < m_num_lo)
+                                ? std::fabs(h_particles_lo.data[lo_idx].y - m_target_momentum)
+                                : std::numeric_limits<Scalar>::max();
+                MPI_Allreduce(MPI_IN_PLACE,
+                              &lo_swap,
+                              1,
+                              MPI_HOOMD_SCALAR_INT,
+                              MPI_MINLOC,
+                              mpi_comm);
+                }
```

Comment:
Two things here:
* You can reduce code duplication if you assign the `MPI_Op` inside the if statement.
* You should not use `min()`. For floating-point values, `min()` is the minimum _positive_ value that you can be represented. You should either use `lowest()` as the value that has no values smaller than it, or in this case, zero since you know all the momenta that have been injected have to be positive.

```suggestion
            // find particle in lo slab from all ranks (p > 0)
            MPI_Op lo_op;
            Scalar_Int lo_swap;
            lo_swap.i = rank;
            if (std::isinf(m_target_momentum))
                {
                lo_op = MPI_MAXLOC;
                lo_swap.s = (lo_idx < m_num_lo) ? h_particles_lo.data[lo_idx].y : Scalar(0.0);
                }
            else
                {
                lo_op = MPI_MINLOC;
                lo_swap.s = (lo_idx < m_num_lo)
                                ? std::fabs(h_particles_lo.data[lo_idx].y - m_target_momentum)
                                : std::numeric_limits<Scalar>::max();
                }
                MPI_Allreduce(MPI_IN_PLACE,
                              &lo_swap,
                              1,
                              MPI_HOOMD_SCALAR_INT,
                              lo_op,
                              mpi_comm);
```